content/5_Plat/Centroid.mdx

```diff
@@ -302,88 +314,151 @@ private static class Centroid {
+vector<bool> is_removed;
+vector<vector<pair<int, int>>> ancestors;
+
+int get_subtree_size(int node, int parent = -1) {
+	subtree_size[node] = 1;
+	for (int child : adj[node]) {
+		if (child == parent || is_removed[child]) { continue; }
+		subtree_size[node] += get_subtree_size(child, node);
+	}
+	return subtree_size[node];
+}
 
-template <typename T> bool ckmin(T &a, const T &b) {
-	return b < a ? a = b, 1 : 0;
+int get_centroid(int node, int tree_size, int parent = -1) {
+	for (int child : adj[node]) {
+		if (child == parent || is_removed[child]) { continue; }
+		if (subtree_size[child] * 2 > tree_size) {
+			return get_centroid(child, tree_size, node);
+		}
+	}
+	return node;
 }
 
-const int MN = 1e5 + 10, INF = 0x3f3f3f3f;
-
-int N, M, s[MN], m[MN][2], t, b, d;
-bool r[MN], red[MN];
-std::vector<int> a[MN], v[MN];
-struct info {
-  public:
-	int n, b, d;
-};
-std::vector<info> g[MN];
-
-int dfs(int n, int p = 0) {
-	s[n] = 1;
-	for (auto x : a[n])
-		if (x != p && !r[x]) s[n] += dfs(x, n);
-	return s[n];
+/**
+ * Calculate the distance between current `node` and the `centroid` it belongs
+ * to. The distances between a node and all its centroid ancestors are stored
+ * in the vector `ancestors`.
+ * @param cur_dist the distance between `node` and `centroid`
+ */
+void get_dists(int node, int centroid, int parent = -1, int cur_dist = 1) {
+	for (int child : adj[node]) {
+		if (child == parent || is_removed[child]) { continue; }
+		++cur_dist;
+		get_dists(child, centroid, node, cur_dist);
+		--cur_dist;
+	}
+	ancestors[node].push_back({centroid, cur_dist});
 }
-int find(int n, int ms, int p = 0) {
-	for (auto x : a[n])
-		if (!r[x] && x != p && s[x] * 2 > ms) return find(x, ms, n);
-	return n;
+
+void build_centroid_decomp(int node = 0) {
+	int centroid = get_centroid(node, get_subtree_size(node));
+
+	/*
+	 * For all nodes in the subtree rooted at `centroid`, calculate their
+	 * distances to centroid
+	 */
+	for (int child : adj[centroid]) {
+		if (is_removed[child]) { continue; }
+		get_dists(child, centroid, centroid);
+	}
+
+	is_removed[centroid] = true;
+	for (int child : adj[centroid]) {
+		if (is_removed[child]) { continue; }
+		// build the centroid decomposition for all child components
+		build_centroid_decomp(child);
+	}
 }
-void dfs2(int n, int p = 0) {
-	for (auto x : a[n])
-		if (!r[x] && x != p) ++d, dfs2(x, n), --d;
-	g[n].push_back({t, b, d});
+
+/**
+ * Paint `node` red by updating all of its ancestors' minimal distances
+ * to a red node
+ */
+void paint(int node) {
+	for (auto &[ancestor, dist] : ancestors[node]) {
+		min_dist[ancestor] = min(min_dist[ancestor], dist);
+	}
+	min_dist[node] = 0;
 }
-void centroid(int n = 1) {
-	n = find(n, dfs(n));
-	v[n].reserve(a[n].size());
-	for (auto x : a[n])
-		if (!r[x]) {
-			t = n, b = v[n].size(), d = 1;
-			dfs2(x, n);
-			v[n].push_back(INF);
-		}
-	r[n] = 1;
-	for (auto x : a[n])
-		if (!r[x]) centroid(x);
+
+/** Print the minimal distance between `node` to a red node */
```

Comment:
```suggestion
// Print the minimal distance between `node` to a red node
```

Reply:
^
all doc comments should be like that
(not sure about the tildes tho...)